### ISSUE TITLE
ResearchHub In-House Comment/Post Spam Classification Algorithm

### ISSUE THREAD

**Issue by dominikusbrian** [ResearchHub/issues#62]
> If possible it would be nice to get a small subset of past dataset related to this flagged comments/post/thread from ResearchHub and some other that are neutral or considered good post. Something of about 3-5k data entries, or so will do much well in producing test model. 

**Is your feature request related to a problem? Please describe.**
Yes, This feature request is related to the endemic problem of spam inappropriate content such as advetisement, bot post, and other harmful or degradatory content that doesn't belong in ResearchHub. User can flag this and down-votes manually but such is not an effective ways to fend off these bad actors which will eventually only increase in  number and sophistication as the community platform grows. 

Example of the SPAM post are as follows:
![image](https://github.com/ResearchHub/issues/assets/60097047/5c0de758-d533-41bf-86e6-a04c09d65ca9)

![image](https://github.com/ResearchHub/issues/assets/60097047/746395d5-f528-4437-9af3-b5ed065ddad5)

**Describe the solution you'd like**
Instead of using out-of-the box spam filter algo or ML from other providers, RH ideally should attempts toward developing an in-house algorithm for this classification. The algorithm will eventually learn better overtime and be very useful in reducing the damage/issue caused by these bot attacks or spam messages and users. Instead of deleting all these spams, RH could gather the obvious example of real spam posted by malicious user as a Training Data, to optimize a classification model specific for this purpose. 

To begin with, we can start with a simple sentiment analysis algorithm based on nltk sentiment analysis and other general framework but then eventually be refined based on existing labelled data (past deleted/reported comments and post for negative example, and good amount of all other good quality or neutral comments and post). For instance in the past I have developed and deployed algorithm/ model for scam classification based on Discord conversations over the years, primarily using spam and correlation analysis as core framework (See blog post about it [here](https://blog.oceanprotocol.com/decoding-community-engagement-insights-from-the-ocean-protocol-discord-challenge-fc494dc1174a) ) and full report posted on GitHub Repo, [here](https://github.com/DreamBrookTech/ocean-data-challenge/blob/main/2_Discord_11March/What_Does_the_Discord_Says_OceanProtocol_CaseStudies.pdf) 

After the base classification model is established, taking it one notch up, we can think about integration of LLM + ML Classification model, for getting state-of-the-art decision making algorithm. 

See details in[ this paper](https://www.researchhub.com/paper/6688870/interpretable-cross-examination-technique-ice-t-using-highly-informative-features-to-boost-llm-performance) I recently peer-reviewed on ResearchHub, called the Interpretable Cross-Examination Technique (ICE-T), that capable of high accuracy and interpretable algorithm for decision making.

![image](https://github.com/ResearchHub/issues/assets/60097047/7a77fbbb-d97f-46d1-93a9-b2f3f5e56a56)

**Describe alternatives you've considered**
Alternative to developing RH's own ML model/ classification algo, is of course to buy off-the-shelf algorithm, but surely they can't perform at its best without contextual retraining with actual RH's past data. Or other options is to double down on human manual check, We will also need to see whether the upcoming KYC system and system upgrade will help in reducing the amount of malicious actor in the platform. But likely they will still further evolve in sophistication as they have been before (e.g., initially just empty post, now some create artificial text and profile as decoy)

**Comment by dominikusbrian** [comment]
Below is a mini data set for the list of currently existing SPAM user that been actively posting and even "doing" peer-review now, for the past two days and counting. 

Maybe in one of the future updates perhaps after Rep V2 is stable, this Ban User feature could be available to Users that have passed certain bar of Reputation level ? More pairs of eyes and fingers to click/label this would definitely help clean up these mess. At the same time will be a valuable datasets for training AI Model that can do the filtering job specifically for ResearchHuh specific context in the future. For instance the pseudo peer-review post of this spammer is inherent feature of spam post on ResearchHub.  That just recently started  appearing. 

In the past they're more simple like just empty post, empty or short comment. Then they evolve into longer form comment, and now peer-review. Something need to be done. 

spam_user.json

`{
"username":["Meds Fedex",
            "Steven Harrisn",
            "Buy Hydromorphone Online End Of Season Sale 2024",
            "Usapaimeds Online",
            "Buy Carisoprodol By Master Card",
            "Wayrightmeds Online",
            "Buy Alprazolam Online Order Xanax On The Web",
            "James Brown",
            "ORDER NOW - Www.Ultromeds.Com",
            "Usa Pharma",
            "Medspharmausa Store",
            "Usamedstorres Online"
            ],
"profile_link":["https://www.researchhub.com/user/1852013/overview",
                "https://www.researchhub.com/user/1790935/overview",
                "https://www.researchhub.com/user/1852084/overview",
                "https://www.researchhub.com/user/1852061/overview",
                "https://www.researchhub.com/user/1852070/overview",
                "https://www.researchhub.com/user/1852008/overview",
                "https://www.researchhub.com/user/1851428/overview",
                "https://www.researchhub.com/user/1850673/overview",
                "https://www.researchhub.com/user/1850487/overview",
                "https://www.researchhub.com/user/1850648/overview",
                "https://www.researchhub.com/user/1850434/overview",
                "https://www.researchhub.com/user/1850530/overview"
                ]
}`

**Comment by dominikusbrian** [comment]
SPAM also happens in peer-review form now 

![image](https://github.com/ResearchHub/issues/assets/60097047/0abd296b-75e8-4b06-ba21-d58e8e189a79)

**Comment by yattias** [comment]
Thanks for opening @dominikusbrian - appreciate your feedback here.

Spam has been an issue of ours for quite some time. With such a small team, we opted for off-shelf ML solutions.
The onset of the new rep algorithm will hopefully help quite a bit with spam mitigation.

**Comment by dominikusbrian** [comment]
> Thanks for opening @dominikusbrian - appreciate your feedback here.
> 
> Spam has been an issue of ours for quite some time. With such a small team, we opted for off-shelf ML solutions. The onset of the new rep algorithm will hopefully help quite a bit with spam mitigation.

Yes @yattias , completely agree off-shelf ML functions will do decently well for filtering general scam/spam like the ads and dirty post above. However, the scam has now often become less obvious and try to disguise as AI generated peer-review, and various other forms of low-quality content submitted in bulk to deceive the system. This one is a bit concerning. 

Would be nice if its possible to work with some subset of an existing partial databse snapshot which contain good and bad stuff related to this scam/spam or bad actor behaviors. Then I can try build some test model and understand the counter-scam/spam strategy that might work for ResearchHub context. 

Let me know if such is possible and would be happy to play around with the dataset. Scrapping from the frontend doesn't seems like a good option and the ResearchHub API seems to be not ready just yet.  So would be nice if sample dataset could be provided for research and development purpose.
